File: src/aizk/graph/raptor.py
import asyncio
import math
from collections.abc import Callable, Iterable
from itertools import batched
from typing import cast

from loguru import logger
from mainboard.profiling import span
from networkx.algorithms.community.modularity_max import greedy_modularity_communities
from networkx.classes import Graph
from patos import FlexModel, sql
from pgvector import HalfVector, Vector
from pydantic import UUID5, Field
from sqlalchemy import Integer, column, delete, or_
from sqlmodel import select

from ..config import settings
from ..ontology import System
from ..serving.embed import Embedder
from ..serving.extract import LLM
from ..store import Community, Entity, Fact
from ..store.identity import User
from ..store.locking import acquire_locks
from ..store.models.tables import EntityClaim, EntityContent, FactClaim, FactContent
from ..store.vector import CosineVector, cosine_distance
from ..types import Scopes
from .ids import entity_id, fact_id
from .models import Node, RaptorReport

_PART_OF = "part_of"

# SQLModel synthesizes table-model keyword constructors outside the static signatures.
_entity_content = cast("Callable[..., EntityContent]", Entity.Content)
_fact_content = cast("Callable[..., FactContent]", Fact.Content)


def to_floats(vector: HalfVector | Vector | list[float] | None) -> list[float]:
    """Materialize a stored embedding as ordinary float values."""
    assert vector is not None
    return vector.to_list() if isinstance(vector, (HalfVector, Vector)) else vector


def cosine(a: list[float], b: list[float]) -> float:
    """Return cosine similarity, or zero when either vector has no magnitude."""
    dot = sum(x * y for x, y in zip(a, b, strict=True))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(x * x for x in b))
    return dot / (norm_a * norm_b) if norm_a and norm_b else 0.0


def modularity_groups(graph: Graph[int]) -> list[list[int]]:
    """Partition one prepared similarity graph, preserving isolated nodes."""
    if graph.number_of_edges() == 0:
        return [[index] for index in graph.nodes]
    with span("raptor_modularity"):
        return [sorted(group) for group in greedy_modularity_communities(graph)]


def redundant_parent(
    parents: list[tuple[Node, list[float]]], vector: list[float], threshold: float
) -> Node | None:
    """Return an already-staged parent whose summary near-duplicates a new one."""
    return next(
        (
            parent
            for parent, parent_vector in parents
            if cosine(parent_vector, vector) >= threshold
        ),
        None,
    )


class RaptorBuilder(FlexModel):
    """Plan a complete RAPTOR generation before its atomic database replacement."""

    scopes: Scopes
    llm: LLM
    embed: Embedder
    contents: list[EntityContent] = Field(default_factory=list)
    claims: list[EntityClaim] = Field(default_factory=list)
    edges: list[FactContent] = Field(default_factory=list)
    edge_claims: list[FactClaim] = Field(default_factory=list)

    def claim(self, content: EntityContent, level: int, summary: str) -> EntityClaim:
        """Build one exact-scope summary claim."""
        return Entity.Claim(
            content_id=content.id,
            created_by=settings.system_user_id,
            scopes=sorted(self.scopes),
            attributes={"level": level, "summary": summary},
        )

    def leaves(self, communities: list[Community]) -> list[Node]:
        """Stage one level-zero summary entity for every community."""
        nodes: list[Node] = []
        for community in communities:
            content = _entity_content(
                id=entity_id(community.label, System.Entity.RAPTOR_SUMMARY),
                name=community.label,
                type=System.Entity.RAPTOR_SUMMARY,
                embedding=community.embedding,
            )
            claim = self.claim(content, 0, community.summary)
            claim.attributes["community"] = str(community.id)
            self.contents.append(content)
            self.claims.append(claim)
            nodes.append(
                Node(
                    entity_id=content.id,
                    label=community.label,
                    summary=community.summary,
                    embedding=to_floats(community.embedding),
                )
            )
        return nodes

    def connect(self, members: list[Node], parent: Node) -> None:
        """Stage part-of fact content and claims from each child to one parent."""
        for member in members:
            edge = _fact_content(
                id=fact_id(
                    member.entity_id,
                    _PART_OF,
                    parent.entity_id,
                    f"is part of {parent.label}",
                ),
                subject_id=member.entity_id,
                object_id=parent.entity_id,
                predicate=_PART_OF,
                statement=f"is part of {parent.label}",
            )
            self.edges.append(edge)
            self.edge_claims.append(
                Fact.Claim(
                    content_id=edge.id,
                    created_by=settings.system_user_id,
                    scopes=sorted(self.scopes),
                )
            )

    async def similarity_groups(self, embeddings: list[list[float]]) -> list[list[int]]:
        """Build a similarity graph with PostgreSQL vector comparisons."""
        graph: Graph[int] = Graph()
        graph.add_nodes_from(range(len(embeddings)))
        if len(embeddings) < 2:
            return modularity_groups(graph)
        vectors = sql.relation(
            "raptor_vector",
            (
                column("ordinal", Integer),
                column("embedding", CosineVector(settings.embed_dim)),
            ),
            list(enumerate(embeddings)),
        )
        left = vectors.alias("raptor_left")
        right = vectors.alias("raptor_right")
        distance = cosine_distance(left.c.embedding, right.c.embedding)
        with span("raptor_similarity_query"):
            async with User.system(self.scopes) as session:
                pairs = await session.exec(
                    select(left.c.ordinal, right.c.ordinal)
                    .select_from(left.join(right, left.c.ordinal < right.c.ordinal))
                    .where(distance <= 1.0 - settings.raptor_sim_threshold)
                )
                graph.add_edges_from(cast("Iterable[tuple[int, int]]", pairs))
        return modularity_groups(graph)

    async def parent(
        self,
        members: list[Node],
        level: int,
        parents: list[tuple[Node, list[float]]],
    ) -> tuple[Node, bool]:
        """Summarize one cluster and stage a new parent unless this level already has it."""
        with span("raptor_summary"):
            report = await self.llm.generate(
                settings.raptor_rollup_system,
                "Child summaries:\n"
                + "\n".join(
                    f"- {member.label}: {member.summary[: settings.raptor_child_summary_chars]}"
                    for member in members
                ),
                RaptorReport,
            )
        with span("raptor_embedding"):
            [vector] = await self.embed.embed([report.summary], mode="document")
        parent = redundant_parent(parents, vector, settings.raptor_redundancy_threshold)
        created = parent is None
        if parent is None:
            content = _entity_content(
                id=entity_id(report.label, System.Entity.RAPTOR_SUMMARY),
                name=report.label,
                type=System.Entity.RAPTOR_SUMMARY,
                embedding=vector,
            )
            parent = Node(
                entity_id=content.id,
                label=report.label,
                summary=report.summary,
                embedding=vector,
            )
            self.contents.append(content)
            self.claims.append(self.claim(content, level, report.summary))
            parents.append((parent, vector))
        self.connect(members, parent)
        return parent, created

    async def level(
        self, nodes: list[Node], groups: list[list[int]], level: int
    ) -> tuple[list[Node], int]:
        """Plan one tree level and keep reused parents in the next level exactly once."""
        next_nodes: list[Node] = []
        next_ids: set[UUID5] = set()
        parents: list[tuple[Node, list[float]]] = []
        written = 0
        for group_batch in batched(groups, settings.raptor_build_concurrency, strict=False):
            member_groups = [[nodes[index] for index in group] for group in group_batch]
            generated = iter(
                await asyncio.gather(
                    *(
                        self.parent(members, level, parents)
                        for members in member_groups
                        if len(members) > 1
                    )
                )
            )
            for members in member_groups:
                parent, created = (members[0], False) if len(members) == 1 else next(generated)
                if parent.entity_id not in next_ids:
                    next_nodes.append(parent)
                    next_ids.add(parent.entity_id)
                written += created
        return next_nodes, written

    async def build(self, communities: list[Community]) -> int:
        """Plan all levels and return the number of non-leaf summaries staged."""
        if len(communities) < 2:
            return 0
        nodes = self.leaves(communities)
        written = 0
        level = 1
        while len(nodes) > settings.raptor_root_max and level <= settings.raptor_max_levels:
            with span("raptor_clustering"):
                groups = await self.similarity_groups([node.embedding for node in nodes])
                groups = [
                    list(branch)
                    for group in groups
                    for branch in batched(group, settings.raptor_branch_factor, strict=False)
                ]
            if len(groups) >= len(nodes):
                break
            nodes, count = await self.level(nodes, groups, level)
            written += count
            level += 1
        return written

    async def replace(self) -> None:
        """Replace one scope's generation atomically while sharing content across scopes.

        A transaction-scoped advisory lock keyed by the canonical scope serializes
        concurrent builds, and the stale generation is reselected under that lock so a
        racing build can never resurrect or double-delete another generation's rows.
        """
        scope_list = sorted(self.scopes)
        async with User.system().owner as opened:
            await acquire_locks(
                opened,
                ["raptor|" + ",".join(str(scope) for scope in scope_list)],
            )
            stale = list(
                await opened.exec(
                    select(Entity.Claim.content_id)
                    .join(Entity.Content, Entity.Content.id == Entity.Claim.content_id)
                    .where(
                        Entity.Claim.scopes == scope_list,
                        Entity.Content.type == System.Entity.RAPTOR_SUMMARY,
                    )
                )
            )
            stale_edges: list[UUID5] = []
            if stale:
                stale_edges = list(
                    await opened.exec(
                        select(Fact.Claim.content_id)
                        .join(Fact.Content, Fact.Content.id == Fact.Claim.content_id)
                        .where(
                            Fact.Claim.scopes == scope_list,
                            Fact.Content.predicate == _PART_OF,
                            or_(
                                Fact.Content.subject_id.in_(stale),
                                Fact.Content.object_id.in_(stale),
                            ),
                        )
                    )
                )
                if stale_edges:
                    await opened.exec(
                        delete(Fact.Claim).where(
                            Fact.Claim.scopes == scope_list,
                            Fact.Claim.content_id.in_(stale_edges),
                        )
                    )
                await opened.exec(
                    delete(Entity.Claim).where(
                        Entity.Claim.scopes == scope_list,
                        Entity.Claim.content_id.in_(stale),
                    )
                )
            await Entity.Content.mint_all(opened, self.contents)
            opened.add_all(self.claims)
            await opened.flush()
            await Fact.Content.mint_all(opened, self.edges)
            opened.add_all(self.edge_claims)
            await opened.flush()
            if stale_edges:
                claimed = (
                    select(Fact.Claim.id).where(Fact.Claim.content_id == Fact.Content.id).exists()
                )
                await opened.exec(
                    delete(Fact.Content).where(
                        Fact.Content.id.in_(stale_edges),
                        ~claimed,
                    )
                )
            if stale:
                claimed_entity = (
                    select(Entity.Claim.id)
                    .where(Entity.Claim.content_id == Entity.Content.id)
                    .exists()
                )
                await opened.exec(
                    delete(Entity.Content).where(
                        Entity.Content.id.in_(stale),
                        ~claimed_entity,
                    )
                )


async def build_raptor(
    llm: LLM,
    embed: Embedder,
    scopes: Scopes | None = None,
) -> int:
    """Build and atomically replace one exact scope's recursive summary tree."""
    key = frozenset(scopes or (settings.system_user_id,))
    builder = RaptorBuilder(scopes=key, llm=llm, embed=embed)
    with span("raptor_snapshot"):
        async with User.system(key) as session:
            communities = list(
                await session.exec(
                    select(Community).where(
                        Community.embedding.is_not(None),
                        Community.scopes == sorted(key),
                    )
                )
            )
    with span("raptor_planning"):
        written = await builder.build(communities)
    with span("raptor_replacement"):
        await builder.replace()
    logger.info("raptor tree wrote {} summaries in scope {}", written, key)
    return written

File: archive/ts/src/lib/server/graph/raptor.ts
// The recursive summary tree, the TS twin of graph/raptor.py: read one exact scope's
// communities and its stale previous generation as the system identity, plan the whole
// tree in memory (level-zero leaves from communities, greedy-modularity clustering of
// summary embeddings, one LLM rollup per multi-member cluster with near-duplicate parents
// reused), then atomically replace the generation in one owner-role transaction, since
// deleting shared entity content is structural work RLS would block.
import { sql } from 'drizzle-orm';
import { v7 as uuid7 } from 'uuid';

import { entityClaim, entityContent, factClaim, factContent } from '../db/schema';
import { embed, structured } from '../serving';
import { settings } from '../settings';
import { asSystem, bypassRls, cosine, scopeKey, toFloats, uuidArray, type Reporter } from './system';

const RAPTOR_SUMMARY = 'raptor_summary';
const PART_OF = 'part_of';

interface RaptorReport {
	label: string;
	summary: string;
}

const ROLLUP_SCHEMA = {
	name: 'RaptorReport',
	schema: {
		type: 'object',
		properties: {
			label: { type: 'string', description: 'short human-readable name for the broader theme' },
			summary: {
				type: 'string',
				description: 'one paragraph grounded only in the child summaries shown'
			}
		},
		required: ['label', 'summary'],
		additionalProperties: false
	}
};

const report: Reporter<RaptorReport> = (system, user) =>
	structured<RaptorReport>(system, user, ROLLUP_SCHEMA);

interface Node {
	entityId: string;
	label: string;
	summary: string;
	embedding: number[];
}

// Cluster summary embeddings by greedy (CNM) modularity over a similarity graph, the
// networkx greedy_modularity_communities semantics: start singleton, repeatedly merge the
// pair with the largest modularity gain while that gain stays positive, largest first.
export const cluster = (embeddings: number[][], threshold: number): number[][] => {
	const count = embeddings.length;
	const edges: [number, number][] = [];
	for (let left = 0; left < count; left++)
		for (let right = left + 1; right < count; right++)
			if (cosine(embeddings[left], embeddings[right]) >= threshold) edges.push([left, right]);
	if (!edges.length) return Array.from({ length: count }, (_, index) => [index]);
	const m = edges.length;
	const degree = new Array<number>(count).fill(0);
	const members = new Map<number, number[]>(
		Array.from({ length: count }, (_, index) => [index, [index]])
	);
	const between = new Map<number, Map<number, number>>();
	const link = (a: number, b: number, weight: number): void => {
		if (!between.has(a)) between.set(a, new Map());
		between.get(a)!.set(b, (between.get(a)!.get(b) ?? 0) + weight);
	};
	for (const [a, b] of edges) {
		degree[a] += 1;
		degree[b] += 1;
		link(a, b, 1);
		link(b, a, 1);
	}
	const totals = new Map<number, number>(degree.map((value, index) => [index, value]));
	for (;;) {
		let best: [number, number] | null = null;
		let bestGain = -Infinity;
		for (const [a, neighbors] of between)
			for (const [b, weight] of neighbors) {
				if (a >= b) continue;
				const gain = weight / m - (totals.get(a)! * totals.get(b)!) / (2 * m * (2 * m));
				if (gain > bestGain) {
					best = [a, b];
					bestGain = gain;
				}
			}
		// networkx merges while the best gain stays non-negative and stops at the first drop.
		if (!best || bestGain < 0) break;
		const [into, from] = best;
		members.get(into)!.push(...members.get(from)!);
		members.delete(from);
		totals.set(into, totals.get(into)! + totals.get(from)!);
		totals.delete(from);
		for (const [neighbor, weight] of between.get(from)!) {
			between.get(neighbor)!.delete(from);
			if (neighbor === into) continue;
			link(into, neighbor, weight);
			link(neighbor, into, weight);
		}
		between.delete(from);
		between.get(into)!.delete(from);
	}
	return [...members.values()]
		.map((group) => [...group].sort((a, b) => a - b))
		.sort((a, b) => b.length - a.length || a[0] - b[0]);
};

const redundantParent = (
	parents: [Node, number[]][],
	vector: number[],
	threshold: number
): Node | null =>
	parents.find(([, parentVector]) => cosine(parentVector, vector) >= threshold)?.[0] ?? null;

interface CommunityRow {
	id: string;
	label: string;
	summary: string;
	embedding: string;
}

// Plan a complete RAPTOR generation before its atomic database replacement, the
// RaptorBuilder twin holding staged contents, claims, edges, and edge claims.
class RaptorBuilder {
	contents: { id: string; name: string; type: string; embedding: string }[] = [];
	claims: {
		id: string;
		contentId: string;
		createdBy: string;
		scopes: string[];
		attributes: Record<string, unknown>;
	}[] = [];
	edges: {
		id: string;
		subjectId: string;
		objectId: string;
		predicate: string;
		statement: string;
	}[] = [];
	edgeClaims: { id: string; contentId: string; createdBy: string; scopes: string[] }[] = [];

	constructor(
		readonly key: string[],
		readonly rollup: Reporter<RaptorReport>
	) {}

	claim(contentId: string, level: number, summary: string): (typeof this.claims)[number] {
		return {
			id: uuid7(),
			contentId,
			createdBy: settings.systemUserId,
			scopes: this.key,
			attributes: { level, summary }
		};
	}

	leaves(communities: CommunityRow[]): Node[] {
		return communities.map((row) => {
			const contentId = uuid7();
			this.contents.push({
				id: contentId,
				name: row.label,
				type: RAPTOR_SUMMARY,
				embedding: row.embedding
			});
			const claim = this.claim(contentId, 0, row.summary);
			claim.attributes.community = row.id;
			this.claims.push(claim);
			return { entityId: contentId, label: row.label, summary: row.summary, embedding: toFloats(row.embedding) };
		});
	}

	connect(members: Node[], parent: Node): void {
		for (const member of members) {
			const edge = {
				id: uuid7(),
				subjectId: member.entityId,
				objectId: parent.entityId,
				predicate: PART_OF,
				statement: `is part of ${parent.label}`
			};
			this.edges.push(edge);
			this.edgeClaims.push({
				id: uuid7(),
				contentId: edge.id,
				createdBy: settings.systemUserId,
				scopes: this.key
			});
		}
	}

	async parent(
		members: Node[],
		level: number,
		parents: [Node, number[]][]
	): Promise<[Node, boolean]> {
		const rollup = await this.rollup(
			settings.raptorRollupSystem,
			'Child summaries:\n' + members.map((member) => `- ${member.label}: ${member.summary}`).join('\n')
		);
		const [literal] = await embed([rollup.summary], 'document');
		const vector = toFloats(literal);
		let parent = redundantParent(parents, vector, settings.raptorRedundancyThreshold);
		const created = parent === null;
		if (parent === null) {
			const contentId = uuid7();
			this.contents.push({
				id: contentId,
				name: rollup.label,
				type: RAPTOR_SUMMARY,
				embedding: literal
			});
			parent = { entityId: contentId, label: rollup.label, summary: rollup.summary, embedding: vector };
			this.claims.push(this.claim(contentId, level, rollup.summary));
			parents.push([parent, vector]);
		}
		this.connect(members, parent);
		return [parent, created];
	}

	async level(nodes: Node[], groups: number[][], level: number): Promise<[Node[], number]> {
		const nextNodes: Node[] = [];
		const nextIds = new Set<string>();
		const parents: [Node, number[]][] = [];
		let written = 0;
		for (const group of groups) {
			const members = group.map((index) => nodes[index]);
			const [parent, created] =
				members.length === 1 ? ([members[0], false] as [Node, boolean]) : await this.parent(members, level, parents);
			if (!nextIds.has(parent.entityId)) {
				nextNodes.push(parent);
				nextIds.add(parent.entityId);
			}
			written += created ? 1 : 0;
		}
		return [nextNodes, written];
	}

	async build(communities: CommunityRow[]): Promise<number> {
		if (communities.length < 2) return 0;
		let nodes = this.leaves(communities);
		let written = 0;
		let level = 1;
		while (nodes.length > settings.raptorRootMax && level <= settings.raptorMaxLevels) {
			const groups = cluster(nodes.map((node) => node.embedding), settings.raptorSimThreshold);
			if (groups.length >= nodes.length) break;
			const [nextNodes, count] = await this.level(nodes, groups, level);
			nodes = nextNodes;
			written += count;
			level += 1;
		}
		return written;
	}

	// Atomically delete the stale generation and insert the complete staged one.
	async replace(stale: string[]): Promise<void> {
		await bypassRls(async (tx) => {
			if (stale.length)
				await tx.execute(sql`delete from entity_content where id = any(${uuidArray(stale)}::uuid[])`);
			if (this.contents.length) await tx.insert(entityContent).values(this.contents);
			if (this.claims.length) await tx.insert(entityClaim).values(this.claims);
			if (this.edges.length) await tx.insert(factContent).values(this.edges);
			if (this.edgeClaims.length) await tx.insert(factClaim).values(this.edgeClaims);
		});
	}
}

// Build and atomically replace one exact scope's recursive summary tree, the build_raptor
// twin, returning the number of non-leaf summaries written.
export const buildRaptor = async (
	scopes?: string[],
	rollup: Reporter<RaptorReport> = report
): Promise<number> => {
	const key = scopeKey(scopes);
	const { communities, stale } = await asSystem(key, async (tx) => {
		const communities = (await tx.execute(sql`
			select id, label, summary, embedding from community
			where embedding is not null and scopes = ${uuidArray(key)}::uuid[]
		`)) as unknown as CommunityRow[];
		const stale = (await tx.execute(sql`
			select ec.content_id from entity_claim ec
			join entity_content e on e.id = ec.content_id
			where ec.scopes = ${uuidArray(key)}::uuid[] and e.type = ${RAPTOR_SUMMARY}
		`)) as unknown as { content_id: string }[];
		return { communities, stale: stale.map((row) => row.content_id) };
	});
	const builder = new RaptorBuilder(key, rollup);
	const written = await builder.build(communities);
	await builder.replace(stale);
	return written;
};
